cli/src/cmd/cast/storage.rs

```diff
@@ -0,0 +1,256 @@
+
+    // Forge
+    #[clap(flatten)]
+    build: build::CoreBuildArgs,
+}
+
+impl StorageArgs {
+    pub async fn run(self) -> Result<()> {
+        let Self { address, block, build, rpc_url, slot, chain, etherscan_api_key } = self;
+
+        let rpc_url = try_consume_config_rpc_url(rpc_url)?;
+        let provider = try_get_http_provider(rpc_url)?;
+
+        let address = match address {
+            NameOrAddress::Name(name) => provider.resolve_name(&name).await?,
+            NameOrAddress::Address(address) => address,
+        };
+
+        // Slot was provided, perform a simple RPC call
+        if let Some(slot) = slot {
+            let cast = Cast::new(provider);
+            println!("{}", cast.storage(address, slot, block).await?);
+            return Ok(())
+        }
+
+        // No slot was provided
+        // Get deployed bytecode at given address
+        let address_code = provider.get_code(address, block).await?;
+        if address_code.is_empty() {
+            eyre::bail!("Provided address has no deployed code and thus no storage");
+        }
+
+        // Check if we're in a forge project
+        let mut project = build.project()?;
+        if project.paths.has_input_files() {
+            // Find in artifacts and pretty print
+            add_storage_layout_output(&mut project);
+            let out = compile(&project, false, false)?;
+            let match_code = |artifact: &ConfigurableContractArtifact| -> Option<bool> {
+                let bytes =
+                    artifact.deployed_bytecode.as_ref()?.bytecode.as_ref()?.object.as_bytes()?;
+                Some(bytes == &address_code)
+            };
+            let artifact =
+                out.artifacts().find(|(_, artifact)| match_code(artifact).unwrap_or_default());
+            if let Some((_, artifact)) = artifact {
+                return fetch_and_print_storage(provider, address, artifact, true).await
+            }
+        }
+
+        // Not a forge project or artifact not found
+        // Get code from Etherscan
+        println!("No matching artifacts found, fetching source code from Etherscan...");
+        let api_key = etherscan_api_key.or_else(|| {
+            let config = Config::load();
+            config.get_etherscan_api_key(Some(chain))
+        }).wrap_err("No Etherscan API Key is set. Consider using the ETHERSCAN_API_KEY env var, or setting the -e CLI argument or etherscan-api-key in foundry.toml")?;
+        let client = Client::new(chain, api_key)?;
+        let source = find_source(client, address).await?;
+        let metadata = source.items.first().unwrap();
+        if metadata.is_vyper() {
+            eyre::bail!("Contract at provided address is not a valid Solidity contract")
+        }
+
+        let version = metadata.compiler_version()?;
+        let auto_detect = version < MIN_SOLC;
+
+        // Create a new temp project
+        // TODO: Cache instead of using a temp directory: metadata from Etherscan won't change
```

Comment:
this isn't very important and we shouldn't cache this in .foundry imo

Reply:
yeah its more so you dont have to download+compile when making multiple queries on the same address